### PR TITLE
Associate Import and Create Exports in Batches

### LINKED ISSUE #2870: [Feature]: Performance Improvements
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe.

Let's speed up the Grouparoo Application!

### Describe the solution you'd like.

_No response_

### Describe an alternate solution.

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by evantahler**
This PR takes two of our slowest tasks, `import:associateRecord` and `record:export` and replaces them with bulk tasks that operate on many records at once.  We trade many slow tasks for a few big tasks, with big SQL reads and writes.   You'll notice a lot more in-memory finding/mapping/filtering of a big batch of records and properties - this is faster than lots of small SQL statements to find records and properties one-by-one for reach record being considered.

This PR also uses caching and bulk writes a little better.

### Development

- [x] Application changes have been tested appropriately

### Impact

- [x] Code follows company security practices and guidelines
- [x] Security impact of change has been considered
- [x] Performance impact of change has been considered
- [x] Possible migration needs considered (model migrations, config migrations, etc.)

Please explain any security, performance, migration, or other impacts if relevant:

### Code review

- [x] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached where applicable.
- [x] Relevant tags have been added to the PR (bug, enhancement, internal, etc.)

### CODE REVIEW

File: core/src/modules/ops/import.ts
Comment (evantahler) [inline core/src/modules/ops/import.ts]: Now that this method will create many records in bulk, and will be run by a unique, timed task, we can not worry about the locks and retries.  Postgres' (and Sqlite's) `updateOnDuplicate` does what we want

File: core/src/modules/ops/record.ts
Comment (evantahler) [inline core/src/modules/ops/record.ts:799]: Similarly to the above, we don't need the lock anymore

File: core/src/models/Import.ts
Comment (evantahler) [inline core/src/models/Import.ts:171]: Association can now only be done by the Op or task

File: core/src/modules/ops/record.ts
Comment (evantahler) [inline core/src/modules/ops/record.ts]: This section of the method adds an extra SQL select to find all the record properties that are unique and have the same rawValues.  Previously, we were relying on `updateOnDuplicate` to handle this automatically, but that had the side-effect of ending the SQL transaction, which makes doing anything after this method not possible in a CLSAction or CLSTask.  

This new query might be slow, so we will need to test it

**Comment by evantahler** [comment]
Running `timed-run` action after latest commit [](https://github.com/evantahler)@evantahler
[consider batch sizes more often](https://github.com/grouparoo/grouparoo/pull/2968/commits/bc93b32ba1565b6943454af8d92c77be870c2a16)


https://github.com/grouparoo/grouparoo/runs/5297765816

**Comment by evantahler** [comment]
Well, this is promising!  Looking like an overall 30% improvement!

Screen Shot 2022-02-23 at 8 21 51 AM

Things do appear to be getting better!   The downside is that now we've lost the ability to trace our bigger tasks in Sentry

File: core/src/modules/ops/group.ts
Comment (evantahler) [inline core/src/modules/ops/group.ts:115]: In a few places we are now trying to respect `config.batchSize.internalWrite` more by only updating 1000 records/rows at a time

File: core/src/tasks/record/makeReady.ts
Comment (evantahler) [inline core/src/tasks/record/makeReady.ts:28]: To skip waiting for the next task in ~10s, we can immediately enqueue a new instance of the task if we got records

File: plugins/@grouparoo/sentry/src/initializers/sentry.ts
Comment (evantahler) [inline plugins/@grouparoo/sentry/src/initializers/sentry.ts]: Now that we have these giant bulk update queries, sentry was dropping larger event payloads.  This truncates our SQL statement lengths so we get *something* at least

File: core/src/modules/ops/record.ts
Comment (evantahler) [inline core/src/modules/ops/record.ts:774]: This method is drastically updated to do *everything* in batches now.  It never throws, but returns a response payload with per-record errors if needed

File: core/src/modules/ops/record.ts
Comment (evantahler) [inline core/src/modules/ops/record.ts:403]: As the comment says, we can't rely on `updateOnDuplicate` throwing like we used to as we need this method to succeed in-line

**Comment by evantahler** [comment]
Ok, this is finally ready for review! 🥳 

Final timed runs 
- https://github.com/grouparoo/grouparoo/runs/5326416971?check_suite_focus=true -> [Sentry](https://sentry.io/organizations/grouparoo/performance/?breakdown=db&display=latency&end=2022-02-25T20%3A06%3A33&project=5807413&start=2022-02-25T15%3A09%3A40&unselectedSeries=p100%28%29)
- https://github.com/grouparoo/grouparoo/runs/5339344316?check_suite_focus=true -> [Sentry](https://sentry.io/organizations/grouparoo/performance/?breakdown=none&end=2022-02-26T05%3A35%3A16&project=5807413&query=transaction.duration%3A%3C15m&showTransactions=p95&sort=-p50&start=2022-02-25T20%3A04%3A30)

---

* [[TEST] revert batch for makeExports](https://github.com/grouparoo/grouparoo/pull/2968/commits/c2c869bcec626504dab1c46ddfd42f29b44878f6) | https://github.com/grouparoo/grouparoo/runs/5380958642?check_suite_focus=true -> [Sentry](). --- Stopped and got stuck
* [consistenet order for recordProperties needed](https://github.com/grouparoo/grouparoo/pull/2968/commits/987ef747570f33f5864a3aaaf2b5f47235f881b5) | https://github.com/grouparoo/grouparoo/runs/5383157733?check_suite_focus=true -> Sentry

File: core/src/modules/ops/destination.ts
Comment (pedroslopez) [inline core/src/modules/ops/destination.ts]: should this be looking for `complete` exports? This was the previous behavior and I believe the correct one to use for pulling old properties and groups

File: core/src/modules/ops/import.ts
Comment (pedroslopez) [inline core/src/modules/ops/import.ts:119]: Note: The previous `associateRecord` task used to also update the record with the properties that came in the import data. This is now happening within the `findOrCreateByUniqueRecordProperties` method.

File: core/src/modules/ops/record.ts
Comment (pedroslopez) [inline core/src/modules/ops/record.ts]: is this `i` being used anywhere?

File: core/src/modules/ops/record.ts
Comment (pedroslopez) [inline core/src/modules/ops/record.ts:944]: 👍🏼 `addOrUpdateProperties` is now happening for records that were found

File: core/src/modules/ops/record.ts
Comment (pedroslopez) [inline core/src/modules/ops/record.ts]: Slightly unrelated since the previous implementation seems to be doing the same thing: does this mean that when calling `addOrUpdateProperties` with a partial hash, we lose the value of the other properties that were not included in the hash?

I'm seeing the same happen in the old code but I feel like I'm missing something since I don't think the values get cleared when associating imports to existing records🤔

**Review by pedroslopez — COMMENTED** [review]
Nice, this is a big one! 

Looks like tests are failing after the option validation merge due to a missing method on the test plugin.

**Comment by evantahler** [comment]
[987ef74](https://github.com/grouparoo/grouparoo/pull/2968/commits/987ef747570f33f5864a3aaaf2b5f47235f881b5) fixes the problem outlined in https://github.com/grouparoo/grouparoo/pull/3007#issuecomment-1055925450

**Comment by evantahler** [comment]
@pedroslopez I've reverted our test [[TEST] revert batch for makeExports](https://github.com/grouparoo/grouparoo/pull/2968/commits/c2c869bcec626504dab1c46ddfd42f29b44878f6).  I did notice that the work done in https://github.com/grouparoo/grouparoo/pull/3007#issuecomment-1055925450 introduced a bug, and potentially was leading to lots of slow behavior, and sometimes getting 'stuck', which has been fixed.

While there's certainly more tuning that can be done, I think that it's time to merge this in and unblock the new work.

File: core/src/modules/ops/recordProperty.ts
Comment (pedroslopez) [inline core/src/modules/ops/recordProperty.ts:159]: Noticed that the old one ordered by `createdAt` - this should be ok because only a consistent order is needed, and `recordId` is faster because of the index, right?

**Review by pedroslopez — APPROVED** [review]

File: core/src/modules/ops/recordProperty.ts
Comment (evantahler) [reply core/src/modules/ops/recordProperty.ts:159]: You got it!

**Pull request merged**
